File: scripts/verify_office_audit_native.ts
import { mkdir, writeFile } from 'node:fs/promises';
import { join, resolve } from 'node:path';
const { DocumentOperationBackend } = require(resolve(process.cwd(), 'build/electron/runtime/actions'));
const { fileSource } = require(resolve(process.cwd(), 'build/electron/runtime/context'));
const { configureDesktop, listWindows, runPowerShellJson, closeDesktop } = require(resolve(process.cwd(), 'build/electron/runtime/desktop'));
type Json = Record<string, any>;
type PatchOperation = Json;

async function main() {
  const root = process.cwd(),
    output = join(root, 'data', `acceptance-office-${Date.now()}`),
    results: Record<string, Json> = {};
  configureDesktop(root);
  await mkdir(output, { recursive: true });
  const script = (data: unknown, body: string) =>
    `$p=[Text.Encoding]::UTF8.GetString([Convert]::FromBase64String('${Buffer.from(JSON.stringify(data)).toString('base64')}'))|ConvertFrom-Json\n${body}`;
  for (const app of ['word', 'excel', 'powerpoint']) {
    const path = join(
      output,
      `${app}.${app === 'word' ? 'docx' : app === 'excel' ? 'xlsx' : 'pptx'}`,
    );
    let created: Json = {};
    try {
      created = await runPowerShellJson(
        script(
          { app, path },
          `$prog=if($p.app -eq 'word'){'Word.Application'}elseif($p.app -eq 'excel'){'Excel.Application'}else{'PowerPoint.Application'}
$owned=$false;try{$app=[Runtime.InteropServices.Marshal]::GetActiveObject($prog)}catch{$app=New-Object -ComObject $prog;$owned=$true}
if($p.app -eq 'word'){$doc=$app.Documents.Add();$doc.Content.Text='Keep old ending';$doc.Range(0,5).Bold=-1;$doc.Range(5,8).Italic=-1;$doc.Range(8,15).Underline=1;$doc.SaveAs2([string]$p.path);$hwnd=[int64]$doc.Windows.Item(1).Hwnd;@{hwnd=$hwnd;owned=$owned;path=$p.path}|ConvertTo-Json -Compress}
elseif($p.app -eq 'excel'){$doc=$app.Workbooks.Add();$sheet=$doc.Worksheets.Item(1);$sheet.Range('A1').Value2=0;$sheet.Range('B1').Value2=0;$doc.SaveAs([string]$p.path,51);$sheet.Range('A1').Locked=$false;$sheet.Protect();@{hwnd=[int64]$doc.Windows.Item(1).Hwnd;sheet=[string]$sheet.Name;owned=$owned;path=$p.path}|ConvertTo-Json -Compress}
else{$app.Visible=-1;$doc=$app.Presentations.Add(-1);$slide=$doc.Slides.Add(1,12);$shape=$slide.Shapes.AddShape(1,30,40,200,100);$shape.Fill.Visible=0;$shape.Line.Visible=0;$doc.SaveAs([string]$p.path);@{hwnd=[int64]$doc.Windows.Item(1).HWND;slideId=[int]$slide.SlideID;shapeId=[int]$shape.Id;owned=$owned;path=$p.path}|ConvertTo-Json -Compress}`,
        ),
        undefined,
        30000,
      );
      const source = fileSource('native-office-audit', path);
      source.identity.hwnd = created.hwnd;
      if (app === 'powerpoint') {
        const window = (await listWindows()).find(
          (item: Json) =>
            String(item.title).includes('powerpoint') &&
            /powerpnt/i.test(String(item.process_name)),
        );
        if (window) source.identity.hwnd = window.hwnd;
      }
      const backend = new DocumentOperationBackend([source]),
        operation: PatchOperation = {
          operationId: app,
          sourceId: source.sourceId,
          referenceId: app,
          operation:
            app === 'word'
              ? 'replace_text'
              : app === 'excel'
                ? 'set_cell_values'
                : 'set_shape_style',
          locator: {
            kind: app === 'word' ? 'text' : app === 'excel' ? 'cell-range' : 'slide-shape',
            value:
              app === 'word'
                ? { start: 0, end: 15 }
                : app === 'excel'
                  ? { sheet: created.sheet, range: 'A1:B1' }
                  : { slideId: created.slideId, shapeId: created.shapeId },
          },
          before:
            app === 'word'
              ? 'Keep old ending'
              : app === 'excel'
                ? [[0, 0]]
                : { fillRgb: null, lineRgb: null },
          after:
            app === 'word'
              ? 'Keep longer ending'
              : app === 'excel'
                ? [[1, 2]]
                : { fillRgb: 255, lineRgb: null },
        };
      const result = await backend.execute(operation),
        current = await backend.readCurrent(operation);
      if (app === 'word') {
        const formatting = await runPowerShellJson(
          script(
            { path },
            `$app=[Runtime.InteropServices.Marshal]::GetActiveObject('Word.Application');$doc=$null;foreach($win in $app.Windows){if($win.Document.FullName -eq $p.path){$doc=$win.Document;break}};@{text=[string]$doc.Range([int]0,[int]18).Text;formats=@([int]$doc.Range([int]0,[int]5).Bold,[int]$doc.Range([int]5,[int]11).Italic,[int]$doc.Range([int]11,[int]18).Underline)}|ConvertTo-Json -Compress`,
          ),
        );
        results.word = {
          result,
          formatting,
          passed:
            result.ok &&
            formatting.text === 'Keep longer ending' &&
            JSON.stringify(formatting.formats) === '[-1,-1,1]',
        };
      } else if (app === 'excel')
        results.excel = {
          result,
          current,
          passed:
            result.ok === false &&
            result.wrote === true &&
            JSON.stringify(current.value) === '[[1,0]]',
        };
      else {
        const restored = await backend.execute({
          ...operation,
          operationId: 'restore-style',
          before: operation.after,
          after: operation.before,
        });
        results.powerpoint = { result, restored, passed: result.ok && restored.ok };
      }
    } catch (error) {
      results[app] = { passed: false, error: String(error) };
    } finally {
      if (created.path)
        try {
          await runPowerShellJson(
            script(
              { app, path, owned: created.owned },
              `$prog=if($p.app -eq 'word'){'Word.Application'}elseif($p.app -eq 'excel'){'Excel.Application'}else{'PowerPoint.Application'};$app=[Runtime.InteropServices.Marshal]::GetActiveObject($prog);$collection=if($p.app -eq 'word'){$app.Documents}elseif($p.app -eq 'excel'){$app.Workbooks}else{$app.Presentations};$doc=$null;foreach($candidate in $collection){if($candidate.FullName -eq $p.path){$doc=$candidate;break}};if($doc){if($p.app -eq 'powerpoint'){$doc.Saved=-1;$doc.Close()}elseif($p.app -eq 'word'){$save=0;$doc.Close([ref]$save)}else{$doc.Close($false)}};if($p.owned){$app.Quit()};@{ok=$true}|ConvertTo-Json -Compress`,
            ),
          );
        } catch (error) {
          results[app]!.cleanupError = String(error);
        }
      await writeFile(join(output, 'result.json'), JSON.stringify(results, null, 2));
    }
  }
  console.log(JSON.stringify({ output, results }, null, 2));
  process.exitCode = Object.values(results).every((result) => result.passed === true) ? 0 : 1;
}
main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
}).finally(() => closeDesktop());

File: scripts/verify_office_audit_native.py
from __future__ import annotations
import json
import sys
from datetime import datetime
from pathlib import Path
sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
import win32com.client
from app.actions.office_document import PowerShellLiveOfficeGateway
from app.actions.powerpoint import PowerPointComGateway


def application(name):
    try:
        return win32com.client.GetActiveObject(name), False
    except Exception:
        return win32com.client.DispatchEx(name), True


def native_integer(value):
    return int(value() if callable(value) else value)


def main():
    root = Path(__file__).resolve().parents[1] / "data" / ("acceptance-office-" + datetime.now().strftime("%Y%m%d-%H%M%S"))
    root.mkdir(parents=True, exist_ok=False)
    results = {}
    def save_results():
        (root / "result.json").write_text(json.dumps(results, ensure_ascii=False, indent=2), encoding="utf-8")
    def cleanup(application_object, document_object, own, *, powerpoint=False):
        try:
            if document_object is not None:
                if powerpoint:
                    document_object.Saved = True
                    document_object.Close()
                else:
                    document_object.Close(False)
            if application_object is not None and own:
                application_object.Quit()
        except Exception as exc:
            results.setdefault("cleanupErrors", []).append(str(exc))
        save_results()
    app = document = None
    owned = False
    try:
        app, owned = application("Word.Application")
        document = app.Documents.Add()
        document.Content.Text = "Keep old ending"
        document.Range(0, 5).Bold = -1
        document.Range(5, 8).Italic = -1
        document.Range(8, 15).Underline = 1
        path = root / "word-minimal.docx"
        document.SaveAs2(str(path))
        hwnd = int(document.Windows.Item(1).Hwnd)
        result = PowerShellLiveOfficeGateway().replace_word(path=str(path), hwnd=hwnd, start=0, end=15,
            expected_text="Keep old ending", replacement="Keep longer ending")
        actual = str(document.Range(0, 18).Text)
        formats = [int(document.Range(0, 5).Bold), int(document.Range(5, 11).Italic), int(document.Range(11, 18).Underline)]
        results["word"] = {"gateway": dict(result), "text": actual, "formats": formats,
            "passed": result.get("ok") is True and actual == "Keep longer ending" and formats == [-1, -1, 1]}
        document.Save()
    except Exception as exc:
        results["word"] = {"passed": False, "error": str(exc)}
    finally:
        save_results()
        cleanup(app, document, owned)
    print(json.dumps({"word": results["word"]}, ensure_ascii=False), flush=True)

    app = document = None
    owned = False
    try:
        app, owned = application("Excel.Application")
        document = app.Workbooks.Add()
        sheet = document.Worksheets.Item(1)
        sheet.Range("A1:B1").Value2 = ((0, 0),)
        path = root / "excel-partial.xlsx"
        document.SaveAs(str(path), 51)
        sheet.Range("A1").Locked = False
        sheet.Protect()
        hwnd = int(document.Windows.Item(1).Hwnd)
        result = PowerShellLiveOfficeGateway().set_excel(path=str(path), hwnd=hwnd, sheet=str(sheet.Name),
            address="A1:B1", expected=[[0, 0]], after=[[1, 2]])
        actual = [sheet.Range("A1").Value2, sheet.Range("B1").Value2]
        results["excel"] = {"gateway": dict(result), "values": actual,
            "passed": result.get("ok") is False and result.get("wrote") is True and actual == [1, 0]}
        sheet.Unprotect()
        document.Save()
    except Exception as exc:
        results["excel"] = {"passed": False, "error": str(exc)}
    finally:
        save_results()
        cleanup(app, document, owned)
    print(json.dumps({"excel": results["excel"]}, ensure_ascii=False), flush=True)

    app = document = None
    owned = False
    try:
        app, owned = application("PowerPoint.Application")
        document = app.Presentations.Add(True)
        slide = document.Slides.Add(1, 12)
        shape = slide.Shapes.AddShape(1, 30, 40, 200, 100)
        shape.Fill.Visible = 0
        shape.Line.Visible = 0
        path = root / "powerpoint-style.pptx"
        document.SaveAs(str(path))
        gateway = PowerPointComGateway()
        import win32gui
        handles = []
        win32gui.EnumWindows(lambda handle, _: handles.append(handle) if path.stem in win32gui.GetWindowText(handle) else None, None)
        args = None
        for handle in handles:
            candidate = {"path": str(path), "hwnd": handle, "slide_id": native_integer(slide.SlideID), "shape_id": native_integer(shape.Id)}
            try:
                gateway.read_shape(**candidate)
                args = candidate
                break
            except RuntimeError:
                continue
        if args is None:
            raise RuntimeError("No native window was bound to this exact temporary presentation")
        shown = gateway.set_shape_style(**args, expected={"fillRgb": None, "lineRgb": None}, after={"fillRgb": 255, "lineRgb": None})
        visible = int(shape.Fill.Visible)
        hidden = gateway.set_shape_style(**args, expected={"fillRgb": 255, "lineRgb": None}, after={"fillRgb": None, "lineRgb": None})
        results["powerpoint"] = {"show": dict(shown), "hide": dict(hidden), "shownVisible": visible, "hiddenVisible": int(shape.Fill.Visible),
            "passed": shown.get("ok") is True and hidden.get("ok") is True and visible == -1 and int(shape.Fill.Visible) == 0}
        document.Save()
    except Exception as exc:
        results["powerpoint"] = {"passed": False, "error": str(exc)}
    finally:
        save_results()
        cleanup(app, document, owned, powerpoint=True)
    print(json.dumps({"powerpoint": results["powerpoint"]}, ensure_ascii=False), flush=True)
    (root / "result.json").write_text(json.dumps(results, ensure_ascii=False, indent=2), encoding="utf-8")
    print(str(root), flush=True)
    return 0 if all(results.get(name, {}).get("passed") for name in ("word", "excel", "powerpoint")) else 1


if __name__ == "__main__":
    raise SystemExit(main())
